src/vm/eventpipe.cpp

```diff
@@ -732,66 +533,68 @@ void EventPipe::WriteEventInternal(EventPipeEvent &event, EventPipeEventPayload
         NOTHROW;
         GC_NOTRIGGER;
         MODE_ANY;
+        PRECONDITION(s_pConfig != nullptr);
+        PRECONDITION(s_pSessions != nullptr);
     }
     CONTRACTL_END;
 
+    if (s_pConfig == nullptr || s_pSessions == nullptr)
+        return;
+
     // Exit early if the event is not enabled.
     if (!event.IsEnabled())
-    {
         return;
-    }
 
     // Get the current thread;
     Thread *pThread = GetThread();
 
-    if (s_pConfig == NULL)
-    {
-        // We can't procede without a configuration
-        return;
-    }
-    _ASSERTE(s_pSession != NULL);
-
     // If the activity id isn't specified AND we are in a managed thread, pull it from the current thread.
     // If pThread is NULL (we aren't in writing from a managed thread) then pActivityId can be NULL
     if (pActivityId == NULL && pThread != NULL)
     {
         pActivityId = pThread->GetActivityId();
     }
 
-    if (!s_pConfig->RundownEnabled() && s_pBufferManager != NULL)
+    for (EventPipeSessions::Iterator iterator = s_pSessions->Begin();
+         iterator != s_pSessions->End();
+         ++iterator)
     {
-        s_pBufferManager->WriteEvent(pThread, *s_pSession, event, payload, pActivityId, pRelatedActivityId);
-    }
-    else if (s_pConfig->RundownEnabled())
-    {
-        _ASSERTE(pThread != nullptr);
-        BYTE *pData = payload.GetFlatData();
-        if (pData != NULL)
+        EventPipeSession *pSession = *iterator;
+        _ASSERTE(pSession != nullptr);
+        if (pSession == nullptr)
+            continue;
+
+        if (!pSession->RundownEnabled())
```

Comment:
A performance issue to track if it isn't already on the list - I suspect that one session triggering rundown causes those events to spill over into other sessions that happen to have overlapping keywords enabled. Its probably a low pri issue.